cloudca/resource_cloudca_vpn_user.go

```diff
@@ -0,0 +1,112 @@
+package cloudca
+
+import (
+	"fmt"
+
+	"github.com/cloud-ca/go-cloudca"
+	"github.com/cloud-ca/go-cloudca/services/cloudca"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceCloudcaVpnUser() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceCloudcaVpnUserCreate,
+		Read:   resourceCloudcaVpnUserRead,
+		Delete: resourceCloudcaVpnUserDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"environment_id": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "ID of the environment where the vpn should be created",
+			},
+			"username": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Username of the VPN user",
+			},
+			"password": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Password of the VPN user",
+			},
+		},
+	}
+}
+
+func resourceCloudcaVpnUserCreate(d *schema.ResourceData, meta interface{}) error {
+	ccaResources, rerr := getResourcesForEnvironmentID(meta.(*cca.CcaClient), d.Get("environment_id").(string))
+	if rerr != nil {
+		return rerr
+	}
+
+	remoteAccessVpnUser := cloudca.RemoteAccessVpnUser{
+		Username: d.Get("username").(string),
+		Password: d.Get("password").(string),
+	}
+	_, err := ccaResources.RemoteAccessVpnUser.Create(remoteAccessVpnUser)
+	if err != nil {
+		return fmt.Errorf("Error adding VPN user: %s", err)
+	}
+
+	// TODO: When the CMC API actually returns the ID of the created user, use it.
+	// Currently there is no way to do a 'Get' based on the username, and we don't have the ID, so
+	// we have to list all users and then loop through to match the username in order to find the ID.
+	vpnUsers, err := ccaResources.RemoteAccessVpnUser.List()
+	if err != nil {
+		return fmt.Errorf("Error getting the created VPN user ID: %s", err)
+	}
+	var userID string
+	for _, user := range vpnUsers {
+		if user.Username == d.Get("username").(string) {
+			userID = user.Id
+		}
+	}
+	if userID != "" {
+		d.SetId(userID)
+	} else {
+		return fmt.Errorf("Error finding the created VPN user ID: %s", err)
+	}
+	return resourceCloudcaVpnUserRead(d, meta)
+}
+
+func resourceCloudcaVpnUserRead(d *schema.ResourceData, meta interface{}) error {
+	ccaResources, rerr := getResourcesForEnvironmentID(meta.(*cca.CcaClient), d.Get("environment_id").(string))
+	if rerr != nil {
+		return rerr
+	}
+
+	// Get the user based on the ID
+	vpnUser, err := ccaResources.RemoteAccessVpnUser.Get(d.Id())
+	if err != nil {
+		d.SetId("")
```

Comment:
Should we return an error here?

Reply:
I don't believe so.  If I remember correctly, I had to do it this way because I was able to get into a broken terraform state which I was unable to recover from if I returned an error there.  I believe that the `terraform plan` would error when trying to read the state and you would not be able to successfully run a terraform plan if there was a create vpn user action to be done.

Would you like me to test and validate if I can reproduce that again?

Reply:
Can you try again? It seems weird that `resourceCloudcaVpnUserRead` is setting the ID to nil and returning a success if it can't be read

Reply:
@pdube I have tested and confirmed that by returning an error here can create a situation where terraform will be completely broken.

Here are the steps to reproduce a completely broken terraform if this returns an `error` instead of `nil`.

- Create a vpn user using terraform
- Delete the vpn user via the web UI
- Run `terraform plan`
The project will be permanently broken at this point as it will be impossible to move past this error in the `terraform plan` and no `terraform apply` will be possible here after.
```
Error: Error refreshing state: 1 error(s) occurred:

* cloudca_vpn_user.vpn_user: 1 error(s) occurred:

* cloudca_vpn_user.vpn_user: cloudca_vpn_user.vpn_user: VPN User (id=) not found
```

I have not dug into the details here, but I *suspect* this issue is due to the CloudMC API returning a `500` error instead of a `404` when a _get by id_ request fails for a VPN User.  I believe this changes how terraform handles the error and assumes that an internal server error was actually thrown in the backend...

I have added a comment to the code to make it clearer why this code is written as it is.

Reply:
Thanks for testing this out. I will try to reproduce the 500 error on my side to see why CloudMC is returning a 500 instead of 404

Reply:
I just did some testing of the api and the `Get` with a deleted user returns a `404` as expected.  I will try to find the strange `500` error I ran into and let you know...

Reply:
I just found the `500` error that I was trying to remember how I got.

If you create a VPN user and then try to create another VPN user with the same username, you will get a 500 error.
```
[ERROR] Received HTTP status code 500
[ERROR] Error Code: SERVICE_EXECUTION_ERROR, Message: , Context: {}
```